regression: `No loader is configured for ".md" files` after v5 on development with `@mdx-js/rollup`

### Describe the bug

```log
  ✘ [ERROR] No loader is configured for ".md" files: packages/pkg/README.md

    docs/App.tsx:20:15:
      20 │       ? import(`../packages/${params.name}/README.md`)
         ╵                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


✘ [ERROR] No loader is configured for ".md" files: packages/pkg/CHANGELOG.md

    docs/App.tsx:37:15:
      37 │       ? import(`../packages/${params.name}/CHANGELOG.md`)
```

### Reproduction

https://github.com/un-ts/mono-lib-boilerplate

### Steps to reproduce

```sh
yarn
yarn dev
```

### System Info

```shell
System:
    OS: macOS 14.1.2
    CPU: (10) arm64 Apple M1 Max
    Memory: 27.23 GB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.18.2 - ~/Library/Caches/fnm_multishells/76002_1701950982110/bin/node
    Yarn: 4.0.2 - ~/Library/Caches/fnm_multishells/76002_1701950982110/bin/yarn
    npm: 9.8.1 - ~/Library/Caches/fnm_multishells/76002_1701950982110/bin/npm
    bun: 1.0.15 - /opt/homebrew/bin/bun
  Browsers:
    Brave Browser: 120.1.61.100
    Chrome: 120.0.6099.62
    Safari: 17.1.2
  npmPackages:
    @vitejs/plugin-react-swc: ^3.5.0 => 3.5.0 
    vite: ^5.0.6 => 5.0.6
```


### Used Package Manager

yarn

### Logs

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/src/node/optimizer/scan.ts
+++ b/packages/vite/src/node/optimizer/scan.ts
@@ -550,7 +550,6 @@ function esbuildScanPlugin(
       // should be faster than doing it in the catch-all via js
       // they are done after the bare import resolve because a package name
       // may end with these extensions
-
       const setupExternalize = (
         filter: RegExp,
         doExternalize: (path: string) => boolean,

--- a/packages/vite/src/node/optimizer/scan.ts
+++ b/packages/vite/src/node/optimizer/scan.ts
@@ -561,16 +561,6 @@ function esbuildScanPlugin(
             external: doExternalize(path),
           }
         })
-        // onResolve is not called for glob imports.
-        // we need to add that here as well until esbuild calls onResolve for glob imports.
-        // https://github.com/evanw/esbuild/issues/3317
-        build.onLoad({ filter, namespace: 'file' }, () => {
-          const externalOnLoadResult: OnLoadResult = {
-            loader: 'js',
-            contents: 'export default {}',
-          }
-          return externalOnLoadResult
-        })
       }
 
       // css
@@ -645,6 +635,16 @@ function esbuildScanPlugin(
         return {
           loader,
           contents,
+        }
+      })
+
+      // onResolve is not called for glob imports.
+      // we need to add that here as well until esbuild calls onResolve for glob imports.
+      // https://github.com/evanw/esbuild/issues/3317
+      build.onLoad({ filter: /.*/, namespace: 'file' }, () => {
+        return {
+          loader: 'js',
+          contents: 'export default {}',
         }
       })
     },

